### PR TITLE
Add two level transfer operator between non-nested levels

### PR DESCRIPTION
**Pull request by fdrmrc**
The present PR, written with @peterrum, introduces a two-level transfer operator between (possibly) non-nested multigrid levels, called `MGTwoLevelTransferNonNested`. Since many of the tasks are similar to the ones already implemented in `MGTwoLevelTransfer`, a base class `MGTwoLevelTransferBase` has been added. The current implementation assumes continuous and scalar elements.

This transfer is supposed to be used exactly in the same way as the classical `MGTwoLevelTransfer`, as shown in some of the tests, where levels are created out of randomly distorted and refined triangulations. Other strategies may be chosen, as long as each level can described by a deal.II tria.

If it helps the review process, this PR can be split into smaller PRs.

### CODE REVIEW

File: include/deal.II/multigrid/mg_transfer_global_coarsening.h
Comment (luca-heltai) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.h:158]: Can you add a documentation to the class, explaining what the class is doing?

File: include/deal.II/base/mpi_remote_point_evaluation.h
Comment (luca-heltai) [inline include/deal.II/base/mpi_remote_point_evaluation.h:110]: Can you move the implementation to the cc file?

Comment (luca-heltai) [reply include/deal.II/base/mpi_remote_point_evaluation.h:110]: Also, this seems a small, self contained change. Can you make a separate PR for this one?

File: include/deal.II/non_matching/mapping_info.h
Comment (luca-heltai) [inline include/deal.II/non_matching/mapping_info.h:98]: All changes to this file could go in a separate PR.

File: tests/multigrid-global-coarsening/non_nested_multigrid_01.cc
Comment (luca-heltai) [inline tests/multigrid-global-coarsening/non_nested_multigrid_01.cc:3]: Here and below. 2023

**Review by luca-heltai — CHANGES_REQUESTED** [review]
Very nice! I'm not a great expert on the topic, so I'll let someone else comment on this. Many of the small changes can go in separate PRs, and you should squash the changes to two or three commits to make them readale.

**Comment by peterrum** [comment]
@fdrmrc The 3 PRs are merged. Could you update this PR?

File: include/deal.II/non_matching/mapping_info.h
Comment (peterrum) [inline include/deal.II/non_matching/mapping_info.h:474]:
```suggestion
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:3719]: Don't we support plain (non-vectorized) read access in `ConstraintInfo`? I guess this relates to the newly opened issue #15162.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:3744]: Let us avoid touching data twice. First do
```suggestion
      evaluation_point_results.clear();
      evaluation_point_results.reserve(rpe.get_point_ptrs().size() - 1);
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:3760]: and then do
```suggestion
          evaluation_point_results.push_back(result);
```

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:3798]: Is this really the transpose of the function above? This would mean that the `process_and_evaluate` function works asymmetrically, because in case `n_entries` is larger than 1 somewhere, the lengths of `evaluation_points_results` here is different from the `evaluation_points_results` used in `prolongate_and_add`. Note that the index used here is `i`, whereas it is `j` (aka `ptr[i] + ...`) above. I am confused.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (kronbichler) [inline include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:3864]: Related to my comment above, this at least indicates that `enable_inplace_operations_if_possible` in `MGTwoLevelTransferBase` should not be pure virtual and directly expose this functionality. Otherwise, see the other comments.

File: tests/multigrid-global-coarsening/mg_transfer_util.h
Comment (kronbichler) [inline tests/multigrid-global-coarsening/mg_transfer_util.h:232]: Do we really want to write the result? Should we check it? As tests are run on systems with limited disk performance, I would prefer not writing the data.

**Review by kronbichler — COMMENTED** [review]
Nice, this is great work, thanks a lot @fdrmrc (and @peterrum). I particularly like the extensive tests. There are some things that I would like to have fixed, plus some other parts where I added comments to be addressed later.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (peterrum) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:3798]: Let me answer this one. Indeed the implementation in `RPE` not completely symmetric, which is related to the application cases for which each function was written.

The function `evaluate_and_process()` collect for each point the contributions from all cells.

However, `process_and_evaluate()` sends the data at the point to the relevant cells. The same data is sent to all cells. The copying is done here:

https://github.com/dealii/dealii/blob/09cb1410a77d398fd711cbcf0f9cb07e960c323b/include/deal.II/base/mpi_remote_point_evaluation.h#L467-L475

`evaluation_point_results[i] /= n_entries;` followed by the copying is as far as I see the transpose of averaging, which is done during prolongation.

**Comment by luca-heltai** [comment]
Depends on #15167.

**Review by peterrum — COMMENTED** [review]
Don't forget to add a change-log entry.

File: include/deal.II/matrix_free/tensor_product_kernels.h
Comment (peterrum) [inline include/deal.II/matrix_free/tensor_product_kernels.h:3488]: @kronbichler @bergbauer We needed to rename `i` to some other name, since the compiler was warning that there are two `i` shadowing each other.

Comment (fdrmrc) [reply include/deal.II/matrix_free/tensor_product_kernels.h:3488]: Here's the error message: https://github.com/dealii/dealii/actions/runs/4884140277/jobs/8716458324#step:10:2319

Comment (bergbauer) [reply include/deal.II/matrix_free/tensor_product_kernels.h:3488]: Odd that it only complains about this now but yes this should be a different variable.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (fdrmrc) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h]: Fixed, thanks.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (fdrmrc) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h]: Fixed.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (fdrmrc) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h:3601]: It is, modulo a post-processing to extract locally owned dofs. However, this doesn't generalize to the case of multiple components.

**Comment by fdrmrc** [comment]
I addressed some of the comments in the review. As we discussed in person last week, the choice to have `enable_inplace_operations_if_possible` directly in the base class was done in order to have a unified interface.

File: include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h
Comment (fdrmrc) [reply include/deal.II/multigrid/mg_transfer_global_coarsening.templates.h]: Yes, thanks.

**Review by kronbichler — APPROVED** [review]
Thank you for the adjustments, and thanks for working on the feature, this is really cool. I was involved in advising @fdrmrc so I would appreciate an additional look at this before merging.

**Comment by peterrum** [comment]
@luca-heltai We are happy with the state. There are some follow-up tasks we have collected in https://github.com/dealii/dealii/issues/15148. Could you also take another look and see if you are happy now!?

**Pull request merged**
